### PR TITLE
Connection culling

### PR DESCRIPTION
**Pull request by sharksforarms**
Previous work: https://github.com/h2o/h2o/pull/1890 & https://github.com/h2o/h2o/pull/2470

# Overview

This PR introduces "connection culling" as a measure to reduce the opportunities of hitting the max number of connections/file descriptors.

## Background information

Currently, there are some configuration parameters which can limit the number of connections to h2o, namely:

- `max-connections`: max number of connections
- `handshake-timeout`, `http1-request-timeout`, `http1-request-io-timeout`, `http2-idle-timeout`,
  `http3-idle-timeout`: protocol level timeouts

## Connection culling

Under connection pressure (leading up to max connections), start "culling" (closing/removing) idle connections (connections which have had no I/O) after a certain threshold, which have an age.

Situations where this is useful: High load, malicious attacks: resource exhaustion

Difference from `max-connections`:
- When the number of connections reach `max-connections`, h2o will not accept() the connection (hard-limit). Connection culling should happen when the number of connections reach a threshold (soft-limit) ideally, lower than `max-connections`.
- Connections to be closed should be connections of least importance: at the moment, these are idle connections: connections with no I/O, in an idle state, with a minimum age. This age should ideally be lower than the protocol timeout value.

Note: It is still possible to hit the max number of connections/file descriptors with this enabled. This adds priority on new connections while getting rid of older, idle connections.

To configure the behaviour of connection culling, the following configuration
parameters are introduced:
- `soft-connection-limit`: Number of connections needed to trigger culling of excess connections Should be lower than `max-connections`
- `soft-connection-limit.min-age`: Define the minimum age, in seconds of idle connections before it should be considered for removal Should be lower than existing protocol level timeout values such as: `http1-request-io-timeout`, `http1-request-timeout`, `http2-idle-timeout`

Other design considerations:
- Keeping track of active and inactive connections allows us to easily abstract what is "idle" vs "active"
- "Idle connections" are the default for a connection, they become active upon I/O "Active connections" is defined as connections that currently have I/O, or are in a non-idle state, waiting on something to progress.
- Using a linked list allows us to easily iterate from oldest to newest since newer idle/active connections are appended
- "active" and "idle" lists are protocol agnostic as there is no requirement to have these separate/have iteration available on each protocol.

# Code changes

- Added 2 new parameters `soft-connection-limit` & `soft-connection-limit.min-age`
- Moved `h2o_protocol_callbacks_t` callbacks to `h2o_conn_callbacks_t`
- Changed connection state handling (idle, active, shutdown) these are transitioned/stored on separate linked lists instead of a single
-  Added stats to events handler

### CODE REVIEW

File: src/httpclient.c [hunk 1]
```diff
@@ -429,6 +429,7 @@ h2o_httpclient_head_cb on_connect(h2o_httpclient_t *client, const char *errstr,
                                   h2o_httpclient_proceed_req_cb *proceed_req_cb, h2o_httpclient_properties_t *props,
                                   h2o_url_t *origin)
 {
+    usleep(req_delay*1000);
```
Comment (sharksforarms) [inline src/httpclient.c:432]: Not a huge fan of this. Is there a better way of doing this?

I'd like to connect on a network and application level (h2/h3) but not proceed on I/O for a delay. This does not seem to work with `-i` as the function returns with `on_head` immediately.

Comment (sharksforarms) [reply src/httpclient.c]: Found a working alternative for now.

File: lib/http2/connection.c [hunk 2]
```diff
@@ -81,63 +79,58 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
...
-     * final timeout before closing the connections */
-    if (do_close_stragglers && ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
-        ctx->http2._graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
-        h2o_timer_link(ctx->loop, ctx->globalconf->http2.graceful_shutdown_timeout, &ctx->http2._graceful_shutdown_timeout);
-    }
+static int close_idle_connection(h2o_conn_t *_conn)
+{
+    h2o_http2_conn_t *conn = (void *)_conn;
+    enqueue_goaway(conn, H2O_HTTP2_ERROR_NONE, h2o_iovec_init(H2O_STRLIT("idle timeout")));
+    close_connection(conn);
```
Comment (sharksforarms) [inline lib/http2/connection.c:108]: We could act upon this immediately by adding a `h2o_quic_send(&conn->h3.super);` here, else the connection is marked as closed and a timeout will clean it later

**Comment by sharksforarms** [comment]
Looks like the graceful shutdown takes a while to discard the quic connections (requires a full timeout) and times out the test plan, failing the test

File: include/h2o.h [hunk 3]
```diff
@@ -919,6 +890,18 @@ typedef struct st_h2o_conn_callbacks_t {
...
+    /**
+     * returns number of closed idle connections
+     */
+    int (*close_idle_connection)(h2o_conn_t *conn);
+    /**
+     * shutdown of connection is requested (optional)
+     */
+    void (*request_shutdown)(h2o_conn_t *conn);
+    /**
+     * for each request (optional)
```
Comment (kazuho) [inline include/h2o.h:902]: I think this callback is currently mandatory, isn't it? We use it in the status handler as you correctly noticed.

File: include/h2o.h [hunk 4]
```diff
@@ -644,27 +627,23 @@ struct st_h2o_context_t {
...
     int shutdown_requested;
-
+    /**
+     * link-list of h2o_conn_t
+     */
+    h2o_linklist_t _active_conns;
+    /**
+     * link-list of h2o_conn_t
+     */
+    h2o_linklist_t _idle_conns;
```
Comment (kazuho) [inline include/h2o.h:637]: Maybe use an unnamed struct like others?
```suggestion
    struct {
        h2o_linklist_t active, idle;
    }  _conns;
```

File: include/h2o.h [hunk 5]
```diff
@@ -1006,8 +989,26 @@ struct st_h2o_conn_t {
...
+                _node_next = _node->next;                                                                                          \
+                h2o_conn_t *_h2o_conn = H2O_STRUCT_FROM_MEMBER(h2o_conn_t, _conns, _node);                                         \
+                decl_var = (void *)_h2o_conn;                                                                                      \
+                {                                                                                                                  \
+                    block                                                                                                          \
+                }                                                                                                                  \
+            }                                                                                                                      \
+        }                                                                                                                          \
+    } while (0)
```
Comment (kazuho) [inline include/h2o.h:1010]: I know that this macro was in the original WIP work, but honestly speaking, I'm not sure if this abstraction improves maintainability or not. WDYT?

File: include/h2o.h [hunk 6]
```diff
@@ -919,6 +890,18 @@ typedef struct st_h2o_conn_callbacks_t {
      * debug state callback (optional)
      */
     h2o_http2_debug_state_t *(*get_debug_state)(h2o_req_t *req, int hpack_enabled);
+    /**
+     * returns number of closed idle connections
+     */
+    int (*close_idle_connection)(h2o_conn_t *conn);
```
Comment (kazuho) [inline include/h2o.h:896]: In src/main.c, we have been using `int`as a unit for counting number of connections. But for a formal API that deals with number of connections, I wonder if we should use `size_t` instead. Rationale behind the choice is that it is unsigned and the maximum is bound by the size of the memory (Technically speaking, `size_t` might only represent a portion of the memory available to a program, but that's a pedantic argument I think).

File: lib/core/context.c [hunk 7]
```diff
@@ -190,3 +190,28 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
     h2o_time2str_rfc1123(ctx->_timestamp_cache.value->rfc1123, &gmt);
     h2o_time2str_log(ctx->_timestamp_cache.value->log, ctx->_timestamp_cache.tv_at.tv_sec);
 }
+
+int h2o_context_close_idle_connections(h2o_context_t *ctx, int max_connections_to_close, int min_age)
```
Comment (kazuho) [inline lib/core/context.c:194]: Regarding `min_age`, I think we might want to use `uint64_t` representing milliseconds. I think that's our common pattern in the public API of libh2o (see various timeouts in h2o.h).
Regarding the type of `max_connections_to_close`, please see my comment above.

File: lib/core/context.c [hunk 8]
```diff
@@ -190,3 +190,28 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+    if (max_connections_to_close <= 0)
+        return 0;
+
+    int closed = 0;
+    h2o_linklist_t *conn_list[] = {
+        &ctx->_idle_conns,
+    };
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, conn_list, {
+        struct timeval now = h2o_gettimeofday(ctx->loop);
+        if (now.tv_sec - conn->connected_at.tv_sec < min_age)
```
Comment (kazuho) [inline lib/core/context.c:205]: We can use `h2o_timeval_subtract`.

File: lib/core/context.c [hunk 9]
```diff
@@ -190,3 +190,28 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+            continue;
+        if (conn->callbacks->close_idle_connection == NULL) {
```
Comment (kazuho) [inline lib/core/context.c:207]: Assuming that we would require `h2o_conn_t`'s to define this callback, I do not think we need to test and print an error. Our custom is to let the program die when we detect a violation of design contract, under the belief that doing so helps us maintain the code in good shape.

File: lib/core/context.c [hunk 10]
```diff
@@ -172,12 +171,13 @@ void h2o_context_dispose(h2o_context_t *ctx)
...
     ctx->shutdown_requested = 1;
-    if (ctx->globalconf->http1.callbacks.request_shutdown != NULL)
-        ctx->globalconf->http1.callbacks.request_shutdown(ctx);
-    if (ctx->globalconf->http2.callbacks.request_shutdown != NULL)
-        ctx->globalconf->http2.callbacks.request_shutdown(ctx);
-    if (ctx->globalconf->http3.callbacks.request_shutdown != NULL)
-        ctx->globalconf->http3.callbacks.request_shutdown(ctx);
+
+    h2o_linklist_t *conn_list[] = {&ctx->_active_conns, &ctx->_idle_conns};
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, conn_list, {
```
Comment (kazuho) [inline lib/core/context.c:176]: I was initially not a fan of `H2O_CONN_LIST_FOREACH` (I understand that this came from the original WIP PR), but my resistance is weakening.

Rather, I wonder if this macro would become better if we could write like:
```c
H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, {&ctx->active_conns, &ctx->_idle_conns}, {
    ...
});`
to avoid the cost of declaring a variable for naming the linklists.

File: lib/http1.c [hunk 11]
```diff
@@ -1272,16 +1281,12 @@ static h2o_iovec_t log_request_index(h2o_req_t *req)
...
-        struct st_h2o_http1_conn_t *conn = H2O_STRUCT_FROM_MEMBER(struct st_h2o_http1_conn_t, _conns, node);
-        int ret = cb(&conn->req, cbdata);
-        if (ret != 0)
-            return ret;
-    }
+    struct st_h2o_http1_conn_t *conn = (void *)_conn;
+    int ret = cb(&conn->req, cbdata);
+    if (ret != 0)
+        return ret;
     return 0;
```
Comment (kazuho) [inline lib/http1.c:1290]: Maybe this would do?
```suggestion
    return cb(&conn->req, cbdata);
```

File: include/h2o.h [hunk 12]
```diff
@@ -919,6 +890,18 @@ typedef struct st_h2o_conn_callbacks_t {
...
+    /**
+     * shutdown of connection is requested (optional)
+     */
+    void (*request_shutdown)(h2o_conn_t *conn);
```
Comment (kazuho) [inline include/h2o.h:900]: I do not have a strong opinion here, but we can make this a required callback and let http1.c define an empty function. The cost of calling an empty function is often a no-op on x86.

File: lib/http2/connection.c (hunk 2 above)
Comment (kazuho) [reply lib/http2/connection.c:108]: I'm not sure if want to do an immediate close in `close_idle_function`.

The intent of the function is to gracefully reduce amount of resource used, so that there would be more room for new connections. That's why we try to only close connections that are "idle."

To paraphrase, we do not want to close connections that look idle to us, but on which a request is actually inflight from the client. That's because doing so would disrupt the request inflight, and/or lead to the client reestablish a connection to resend the request. That'd be a lose-lose scenario to both ends.

I think the correct behavior here would be to start the graceful shutdown process by sending GOAWAY. The same argument applies to http3.

(and I recognize that we might need three linklists rather than two, because we'd need a separate linklist for connections in the graceful-shutdown state).

**Review by kazuho — COMMENTED** [review]
Thank you for the PR. I am happy to see this turning into shape.

I haven't been able to go through all the bits, but I think I've read and left comments to the API and the generic direction of the callbacks. I think we are on the right track.

File: include/h2o.h (hunk 3 above)
Comment (sharksforarms) [reply include/h2o.h:902]: Yes you're right, I refactored to check for NULL when previously this was not the case.

File: include/h2o.h (hunk 6 above)
Comment (sharksforarms) [reply include/h2o.h:896]: I agree with using better types. Better yet, size_t is unsigned, too.

File: include/h2o.h (hunk 5 above)
Comment (sharksforarms) [reply include/h2o.h:1018]: Follow up here: https://github.com/h2o/h2o/pull/2932#discussion_r798749534

File: lib/http2/connection.c [hunk 13]
```diff
@@ -82,63 +80,58 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
...
-    /* After waiting a second, we still had active connections. If configured, wait one
-     * final timeout before closing the connections */
-    if (do_close_stragglers && ctx->globalconf->http2.graceful_shutdown_timeout > 0) {
-        ctx->http2._graceful_shutdown_timeout.cb = graceful_shutdown_close_stragglers;
-        h2o_timer_link(ctx->loop, ctx->globalconf->http2.graceful_shutdown_timeout, &ctx->http2._graceful_shutdown_timeout);
-    }
+static size_t close_idle_connection(h2o_conn_t *_conn)
+{
+    h2o_linklist_unlink(&_conn->_conns);
+    h2o_linklist_insert(&_conn->ctx->_conns.shutdown, &_conn->_conns);
```
Comment (kazuho) [inline lib/http2/connection.c:108]: I think we can move these two lines into `initiate_graceful_shutdown`? Or to paraphrase, is there a case where we would want to initiate graceful shutdown but not have them added to the shutdown list?

File: lib/http2/connection.c [hunk 14]
```diff
@@ -82,63 +80,58 @@ static void enqueue_goaway(h2o_http2_conn_t *conn, int errnum, h2o_iovec_t addit
...
      * requests can be initiated. After waiting at least one round trip time, the server can send another GOAWAY frame with an
      * updated last stream identifier. This ensures that a connection can be cleanly shut down without losing requests.
      */
-    h2o_linklist_t *node;
 
+    h2o_http2_conn_t *conn = (void *)_conn;
     /* only doit once */
-    if (ctx->http2._graceful_shutdown_timeout.cb != NULL)
+    if (conn->_graceful_shutdown_timeout.cb != NULL)
         return;
```
Comment (kazuho) [inline lib/http2/connection.c:125]: I think we can remove this `if` statement (it originates from #119 but I do not recall why it was necessary), and rather do `assert(conn->_graceful_shutdown_timeout.cb == NULL)` at the beginning of this function.

Rationale is that it is cleaner to have a contract that requires the callback to be called at most only once per each connection. If we are to provide safeguard, we can consult `ctx->shutdown_requested` in `h2o_context_request_shutdown`.

File: lib/http2/connection.c [hunk 15]
```diff
@@ -156,6 +149,17 @@ static void on_idle_timeout(h2o_timer_t *entry)
...
+    if (!h2o_timer_is_linked(&conn->_graceful_shutdown_timeout)) {
+        h2o_linklist_unlink(&conn->super._conns);
+        if (0 == conn->num_streams.priority.open + conn->num_streams.priority.half_closed + conn->num_streams.pull.open +
+                conn->num_streams.pull.half_closed + conn->num_streams.push.open + conn->num_streams.push.half_closed) {
+            // all streams are idle
+            h2o_linklist_insert(&conn->super.ctx->_conns.idle, &conn->super._conns);
+        } else {
+            h2o_linklist_insert(&conn->super.ctx->_conns.active, &conn->super._conns);
+        }
+    }
```
Comment (kazuho) [inline lib/http2/connection.c:161]: I think it might be possible to move this logic into `h2o_http2_stream_set_state`. The function is responsible for maintaining stream counts.

**Review by kazuho — COMMENTED** [review]
Thank you for your continued efforts. I like how this is coming into shape.

File: lib/core/context.c [hunk 16]
```diff
@@ -190,3 +190,30 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+
+    size_t closed = 0;
+
+    // consider connections in shutdown state as closed
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, ({&ctx->_conns.shutdown}), {
+        conn;
+        closed++;
+        if (closed == max_connections_to_close)
+            return closed;
+    });
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/core/context.c:207]: I had the same thought, adding a counter would make this path more efficient

File: src/main.c [hunk 17]
```diff
@@ -3305,6 +3330,29 @@ static void extra_status_jemalloc_cb(void *ctx, const char *stats)
 }
 #endif
 
+static int count_linklist(const h2o_linklist_t *list)
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply src/main.c:3333]: With a separate counter, we no longer need this :+1:

File: lib/http1.c [hunk 18]
```diff
@@ -181,8 +175,13 @@ static void set_req_io_timeout(struct st_h2o_http1_conn_t *conn, uint64_t timeou
...
         h2o_timer_unlink(&conn->_io_timeout_entry);
     conn->_io_timeout_entry.cb = cb;
-    if (cb != NULL)
+    h2o_linklist_unlink(&conn->super._conns);
+    if (cb != NULL) {
         h2o_timer_link(conn->super.ctx->loop, timeout, &conn->_io_timeout_entry);
+        h2o_linklist_insert(&conn->super.ctx->_conns.idle, &conn->super._conns);
+    } else {
+        h2o_linklist_insert(&conn->super.ctx->_conns.active, &conn->super._conns);
+    }
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/http1.c:194]: > cleanup_connection with http1_is_persistent set to true but is yet to receive a single byte

I'm having troubles re-producing this situation via test case, could you expand on this?

File: lib/core/context.c (hunk 16 above)
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/core/context.c:207]: Results in this commit: https://github.com/h2o/h2o/pull/2932/commits/96d31878ef285a3ba06ea38b71f56abbd9ad8cfc

File: lib/http2/connection.c [hunk 19]
```diff
@@ -156,6 +149,17 @@ static void on_idle_timeout(h2o_timer_t *entry)
 
 static void update_idle_timeout(h2o_http2_conn_t *conn)
 {
+    if (!h2o_timer_is_linked(&conn->_graceful_shutdown_timeout)) {
+        h2o_linklist_unlink(&conn->super._conns);
+        if (0 == conn->num_streams.priority.open + conn->num_streams.priority.half_closed + conn->num_streams.pull.open +
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/http2/connection.c]: Another good article: https://calendar.perfplanet.com/2018/http2-prioritization/#firefox

File: lib/http1.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (kazuho) [reply lib/http1.c:193]: Look at `handle_one_body_fragment`. This function calls `clear_timeouts` that invoke `set_req_io_timeout` with the last parameter set to NULL.

The logic here is to disable I/O timeout while the server processes a chunk of request body that has been received (which can take long time when the server is forwarding that chunk to the origin), then reactivate I/O timeout when the `proceed_request` callback is being called.

I believe we have similar logic on the send-side (i.e., on the path we send response).

To summarize, I/O timeout is enabled only when we the client is expected to make the next state change. It is turned off while the request is inflight but the server is the one to trigger the state change.

[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/http1.c:193]: My assumption is that `cleanup_connection` get's called when h2o has finished reading/processing the request & body and can start reading / sending the next request.

> but is yet to receive a single byte

In this case do you mean the response from the origin? Do we currently collect this information? I see `conn->req->res` but maybe I'm mis-understanding.

[the remark this answers is not in the document]
Comment (kazuho) [reply lib/http1.c:203]: > My assumption is that `cleanup_connection` get's called when h2o has finished reading/processing the request & body and can start reading / sending the next request.

That is correct.

>> but is yet to receive a single byte
> In this case do you mean the response from the origin?

No, I mean (yet to receive a single byte of) a request from client. But to be precise, let's discuss each state of a HTTP/1 connection.

Let's look at each state.
* When the client connects to the server using HTTP/1, `io_timeout` is used, and the connection is *not* in idle state. The belief here is that client is connecting because it is going to send a request.
* When there is something inflight to the client, `io_timeout` is used, because we expect the client to make progress within that timeout as well.
* When the ball is at the server side but the server does not have anything to send (waiting for upstream to supply more data), timeout is disabled (i.e., the callback is set to NULL).
* Once all response have been sent and the connection is to persist, we adopt `idle_timeout`, because at this moment, we do not know if the client is going to immediately send a new request on that connection, or if the client is just going to hold onto the connection as it might want to use it in the future. Once we receive a byte, we know that the client is sending a request, hence we adopt `io_timeout`.

The observation behind this design is that "idle" is a moment when the server thinks a *client* might not have a request inflight. That's when we can cull the connection.

**Review by kazuho — COMMENTED** [review]
Hey @sharksforarms. I'm very sorry for the prolonged review.

File: lib/http1.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/http1.c:204]: Thanks for explaining!

I added a check to make sure the input socket is empty before going idle; do you believe we also need to check the `conn->_io_timeout_entry.cb`? In my testing this is non-null when `cleanup_connection` is called  
https://github.com/h2o/h2o/pull/2932/commits/4cd53680dc2faaed55c93442c0a93fa8673b21bd

File: lib/core/context.c [hunk 20]
```diff
@@ -190,3 +190,30 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+            continue;
+        ctx->connection_stats.idle_closed++;
+        conn->callbacks->close_idle_connection(conn);
+        closed++;
+        if (closed == max_connections_to_close)
+            return;
+    });
+}
+
+size_t *get_connection_state_counter(h2o_context_t *ctx, h2o_connection_state_t state)
```
Comment (kazuho) [inline lib/core/context.c:216]: I really like the idea of abstracting the counters. That said, would it make sense to change this to something like `h2o_set_conn_state(h2o_conn_t *conn, h2o_conn_state_t new_state)`, where:
* the counter is automatically decremented for the old state, 
* incremented for the new state, and
* the state is updated to the given state?

I ask this due to the following reasons:
* We see lots of `--*get_connection_state_counter(...)` followed by `++*get_connection_state_counter(...)`, it's easy to miss `++` after `--`.
* I'm not sure if there's a reason to separate state transition (i.e., assignment to `h2o_conn_t::state`) and counter updates, considering that state is a member of `h2o_conn_t` (which is the superclass of all the conn_t's).


FWIW, `h2o_http2_stream_state` is designed like that.

File: lib/http1.c [hunk 21]
```diff
@@ -158,6 +153,14 @@ static void cleanup_connection(struct st_h2o_http1_conn_t *conn)
         /* TODO use lingering close */
         close_connection(conn, 1);
         return;
+    } else {
```
Comment (kazuho) [inline lib/http1.c:156]: I think we do not need to have an else clause here, because the `if` above uses early return.

File: lib/http1.c [hunk 22]
```diff
@@ -769,8 +786,7 @@ static void on_timeout(struct st_h2o_http1_conn_t *conn)
         conn->req.res.reason = "Request Timeout";
     }
 
-    conn->req.http1_is_persistent = 0;
-    close_connection(conn, 1);
+    close_idle_connection(&conn->super);
```
[the remark this answers is not in the document]
Comment (sharksforarms) [reply lib/http1.c:789]: it's the same logic, I only abstracted into a `close_idle_connection` function.

Is the function name causing confusion with the connection state? My thought was closing a "timed out" connection is the same as closing an idle connection.

File: lib/core/context.c (hunk 20 above)
Comment (sharksforarms) [reply lib/core/context.c]: Added some commits, this looks much better and is less error prone. Thanks!

**Comment by sharksforarms** [comment]
@kazuho Thanks for the review! I have address your comments in the commits above

File: lib/core/context.c [hunk 23]
```diff
@@ -190,3 +190,69 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+
+    size_t closed = ctx->_conns.num_conns.shutdown;
+
+    if (closed >= max_connections_to_close)
+        return;
+
+    H2O_CONN_LIST_FOREACH(h2o_conn_t * conn, ({&ctx->_conns.idle}), {
+        struct timeval now = h2o_gettimeofday(ctx->loop);
+        if (h2o_timeval_subtract(&conn->connected_at, &now) < (min_age * 1000))
+            continue;
```
Comment (kazuho) [inline lib/core/context.c:207]: I wonder if we need this logic at all, based on my understanding that we agreed that connections (specifically H1) are not considered as idle when they have been created but have not received any requests.

Though we might want to make sure that view is reflected to the logic that we have in http2 and http3 as well. For the purpose, IIRC h2 and h3 has the concept of largest stream ID being received, or the next stream ID to be received. We can probably use that to determine the state of the connection.

File: lib/core/context.c [hunk 24]
```diff
@@ -190,3 +190,69 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+            return;
+    });
+}
+
+static size_t *get_connection_state_counter(h2o_context_t *ctx, h2o_conn_state_t state)
+{
+    return ctx->_conns.num_conns.counters + (size_t)state;
+}
+
+static void _connection_state_set(h2o_conn_t *conn, h2o_conn_state_t state)
```
Comment (kazuho) [inline lib/core/context.c:221]:
```suggestion
static void set_conn_state(h2o_conn_t *conn, h2o_conn_state_t state)
```
Maybe? Some rationale:
* this is function is internal by its linkage being static, we do not need `_`.
* when naming a static function, it is our convention to start with a verb, omitting the name of the module.

File: lib/core/context.c [hunk 25]
```diff
@@ -190,3 +190,69 @@ void h2o_context_update_timestamp_string_cache(h2o_context_t *ctx)
...
+    --*get_connection_state_counter(conn->ctx, conn->state);
+    h2o_linklist_unlink(&conn->_conns);
+}
+
+void h2o_conn_set_state(h2o_conn_t *conn, h2o_conn_state_t state)
+{
+    --*get_connection_state_counter(conn->ctx, conn->state);
+    conn->state = state;
+    ++*get_connection_state_counter(conn->ctx, conn->state);
+    _connection_state_set(conn, state);
```
Comment (kazuho) [inline lib/core/context.c:257]: This is nothing more than a nitpick, but it might make sense to create two internal functions:
* one that increments the counter and links the connection, and
* another that decrements the counter and unlinks the connection.

Then, we can call the former in `h2o_conn_init_state`, call the latter in `h2o_conn_fin_state`, and call both in `h2o_conn_set_state`.

**Review by kazuho — COMMENTED** [review]
With the changes, the logic looks very clean. I really like the fact that connection management has now become a version-independent concept!!!!

File: lib/core/context.c (hunk 23 above)
Comment (sharksforarms) [reply lib/core/context.c:207]: H1 are considered as `idle` on creation, `active` once they receive data, then (before parsing a request) `idle` again if there's no more data to process.

I'm not sure I follow the stream id method of determining the state of the connection, could you expand on this idea?

Am I on the right track in thinking:
- If received stream_id > max stream id: idle
- if received stream_id < max stream id: active

**Comment by kazuho** [comment]
Thank you for the changes. I think this PR is ready to land with minor changes, so I've opened #3038.

@sharksforarms Could you take a look?

**Comment by sharksforarms** [comment]
Hi @kazuho I took a look at your changes and they make sense to me. Thanks!
